tests/testsuite/update.rs

```diff
@@ -2096,3 +2095,324 @@ fn update_breaking_specific_packages() {
+        .masquerade_as_nightly_cargo(&["update-breaking"])
+        .with_stderr(
+            "\
+[UPDATING] `[..]` index
+",
+        )
+        .run();
+
+    let crate_manifest_after = p.read_file("Cargo.toml");
+    assert_e2e().eq(&crate_manifest_after, crate_manifest);
+
+    let lock_file_after = p.read_file("Cargo.lock");
+    assert_e2e().eq(&lock_file_after, lock_file);
+
+    p.cargo(
+        "update compatible renamed-from non-semver transitive-compatible transitive-incompatible",
+    )
+    .with_stderr(
+        "\
+[UPDATING] `[..]` index
+[LOCKING] 5 packages to latest compatible versions
+[UPDATING] compatible v1.0.0 -> v1.0.1
+[UPDATING] non-semver v1.0.0 -> v1.0.1 (latest: v2.0.0)
+[UPDATING] renamed-from v1.0.0 -> v1.0.1 (latest: v2.0.0)
+[UPDATING] transitive-compatible v1.0.0 -> v1.0.1
+[UPDATING] transitive-incompatible v1.0.0 -> v1.0.1 (latest: v2.0.0)
+",
+    )
+    .run();
+}
+
+#[cargo_test]
+fn update_breaking_without_lock_file() {
+    Package::new("compatible", "1.0.0").publish();
+    Package::new("incompatible", "1.0.0").publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name  =  "foo"
+            version  =  "0.0.1"
+            edition  =  "2015"
+            authors  =  []
+
+            [dependencies]
+            compatible  =  "1.0"  # Comment
+            incompatible  =  "1.0"  # Comment
+        "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    Package::new("compatible", "1.0.1").publish();
+    Package::new("incompatible", "1.0.1").publish();
+
+    Package::new("incompatible", "2.0.0").publish();
+
+    p.cargo("update -Zunstable-options --breaking")
+        .masquerade_as_nightly_cargo(&["update-breaking"])
+        .with_stderr(
+            "\
+[UPDATING] `[..]` index
+[UPGRADING] incompatible ^1.0 -> ^2.0
+[LOCKING] 3 packages to latest compatible versions
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn update_breaking_spec_version() {
+    Package::new("compatible", "1.0.0").publish();
+    Package::new("incompatible", "1.0.0").publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name  =  "foo"
+            version  =  "0.0.1"
+            edition  =  "2015"
+            authors  =  []
+
+            [dependencies]
+            compatible  =  "1.0"  # Comment
+            incompatible  =  "1.0"  # Comment
+        "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("generate-lockfile").run();
+
+    Package::new("compatible", "1.0.1").publish();
+    Package::new("incompatible", "1.0.1").publish();
+
+    Package::new("incompatible", "2.0.0").publish();
+
+    p.cargo("update -Zunstable-options --breaking incompatible@1.0.0")
+        .masquerade_as_nightly_cargo(&["update-breaking"])
+        .with_stderr(
+            // FIXME: This is wrong.
```

Comment:
I found another problem. We're not matching `incompatible` with `incompatible@1.0.0`.

Reply:
But maybe that's not a bug. Strictly speaking, `1.0.0` does not match the version requirement `1.0`.

Reply:
Anyway, I pushed a fix.

Reply:
Think of this feature as a specially named `--force` flag. They identified the entry they want to update so we should.

What we also need a test for is if they specify a compatible version that isn't in the lockfile.

Reply:
I'm actually not using the lock file. This should work whether or not there is a lock file. I'm just checking if the version matches the version req.

Reply:
Thats an implementation detail while I'm talking of user-facing semantics. We should match `spec` behavior with `cargo update` without `--breaking`.

Reply:
Imagine a non-existing `compatible@1.0.10`. When run without a lock file, `cargo update compatible@1.0.10` does not complain. Then, if the same command is run again (now with a lock file), it complains about `compatible@1.0.10` not matching any packages. So it is not idempotent.

Is it OK if `cargo update --breaking compatible@1.0.10` also doesn't complain, whether or not there is a lock file?

If not OK, we'll have to query a `previous_resolve` just for this error message. And it might be a good idea to go back to reusing the `ops::update_lockfile` that the non-breaking update is running (but then having to make changes to `keep`). In any case, can we put that in a separate PR?

Reply:
@epage This may or may not be a task to put on the task list, but it depends on your answer to the question about `cargo update --breaking compatible@1.0.10` not complaining.

Reply:
Let's put it in the task list either way and not block this conversation further.

Reply:
FYI, this is changed in #14140. If there isn't a lock file, `update -Zunstable-options --breaking incompatible@9.0.0` will do this:

```
[UPDATING] `dummy-registry` index
[LOCKING] 3 packages to latest compatible versions
[ADDING] incompatible v1.0.1 (latest: v2.0.0)
```

If there is a lock file, it will do this:

```
[ERROR] package ID specification `incompatible@9.0.0` did not match any packages
Did you mean one of these?

  incompatible@1.0.0
```